src/service/locator.cpp

```diff
@@ -669,38 +680,46 @@ locator_t::on_routing(const std::string& ruid, bool replace) -> streamed<results
 
 void
 locator_t::on_service(const std::string& name, const results::resolve& meta, modes mode) {
+    scoped_attributes_t attributes(*m_log, { attribute::make("service", name) });
+
     if(m_cfg.restricted.count(name)) {
+        COCAINE_LOG_DEBUG(m_log, "ignoring service update due to restrictions");
         return;
     }
 
-    scoped_attributes_t attributes(*m_log, { attribute::make("service", name) });
+    // TODO(@kobolog):
+    //   - Separate snapshotting and remote stream synchronization, like in RGs vs. Routers.
+    //   - Send updates out from outside the critical section.
+    //   - Batch updates into reasonable-sized announces and flush every few seconds.
 
-    auto mapping = m_remotes.synchronize();
+    m_remotes.apply([&](remote_map_t& mapping) {
+        const auto update = results::connect{m_cfg.uuid, {{name, meta}}};
 
-    if(mode == modes::exposed) {
-        if(m_snapshots.count(name) != 0) {
-            COCAINE_LOG_ERROR(m_log, "duplicate service detected");
-            return;
-        }
+        if(mode == modes::exposed) {
+            if(m_snapshots.count(name) != 0) {
+                COCAINE_LOG_ERROR(m_log, "duplicate service detected");
```

Comment:
I think it'll be great if this message would contain the duplicated service name.